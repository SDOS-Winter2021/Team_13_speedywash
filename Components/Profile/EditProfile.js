/**
 * @module
 */
import React, { useEffect, useState } from 'react'
import { Text, View, Button, TextInput } from 'react-native'
import keys from "../../configs/KEYS"
import AsyncStorage from '@react-native-community/async-storage';
import * as firebase from 'firebase';
import { Image, StyleSheet, Alert } from 'react-native';
import { SafeAreaView } from 'react-native';
import { TouchableOpacity } from 'react-native';
import * as ImagePicker from 'expo-image-picker';
import { setValue } from "../../configs/CacheManager";

/**
 * Allows user to edit his/her displayName, email
 * and upload his/her profile image. All the changes are also 
 * commited to database as well
 * @param {Object} obj - An object
 * @param {module:globaltypes.User} obj.currentUser - current user profile
 * @param {function} obj.setCurrentUser - current user setter
 * @param {function} obj.setcurrentView - current view setter
 * @return {View} - React Component View
*/
function EditProfile({ currentUser, setcurrentUser, setcurrentView }) {

    const [avatarName, setavatarname] = useState(currentUser.displayName);
    const [userDetails, setUserDetails] = useState({
        name: currentUser.displayName,
        email: currentUser.email
    });
    // const [img,setImg] = useState(currentUser.photoURL == "" ? require("../../assets/avatar.png") : currentUser.photoURL);
    const [img, setImg] = useState(currentUser.photoURL == "" ? "../../assets/avatar.png" : currentUser.photoURL);
    const [isPicked, setPicked] = useState(currentUser.photoURL == "" ? false : true);

    const pickImage = () => {
        ImagePicker.requestMediaLibraryPermissionsAsync().then((status) => {
            if (status.granted) {
                ImagePicker.launchImageLibraryAsync({
                    mediaTypes: ImagePicker.MediaTypeOptions.Images,
                    allowsEditing: true,
                    aspect: [4, 3],
                    quality: 1
                }).then((result) => {
                    if (!result.cancelled) {
                        setImg(result.uri)
                        fetch(result.uri).then((response) => {
                            response.blob().then((blob) => {
                                let ref = firebase.storage().ref().child(`profile/${currentUser.uid}`)
                                ref.put(blob).then(() => {
                                    firebase.storage().ref(`profile/${currentUser.uid}`).getDownloadURL().then((url) => {
                                        firebase.firestore().collection("users").doc(currentUser.uid)
                                            .update({ photoURL: url })
                                            .then(() => {
                                                Alert.alert("Success", "Image Uploaded Successfully")
                                                setcurrentUser({ ...currentUser, photoURL: url })
                                            })
                                    })
                                })
                            })
                        })
                        setPicked(true);
                    }
                })
            }
        });
    }
    function handleName(text) {
        setUserDetails(prevValue => {
            return {
                name: text,
                email: prevValue.email
            };
        });
    }
    function handleEmail(text) {
        setUserDetails(prevValue => {
            return {
                name: prevValue.name,
                email: text
            };
        });

    }
    function handleUpdates() {
        const newObj = {
            "displayName": userDetails.name,
            "email": userDetails.email,
        }
        firebase.firestore().collection("users").doc(currentUser.uid).update(newObj).then(() => {

            AsyncStorage.setItem(keys.storage.USER, JSON.stringify({ ...currentUser, ...newObj }))
                .then(() => {
                    setcurrentUser({ ...currentUser, ...newObj })
                    setavatarname(userDetails.name);
                    Alert.alert("Success", "Details Updated Successfully")
                })
        });
    }
    // console.log(img)
    function AvatarCard() {
        return (
            <View style={styles.card}>
                <View>
                    <Text style={styles.cardText}>
                        {avatarName}
                    </Text>
                    <Text style={styles.cardText}>
                        {currentUser.phoneNumber}
                    </Text>

                </View>
                {/* <Image resize Mode="contain" style={styles.cardImage} source={
                    {uri: currentUser.photoURL == "" ? require("../../assets/avatar.png") : currentUser.photoURL}
                    } /> */}
                {isPicked ? <Image resize Mode="contain" style={styles.cardImage} source={{ uri: img }} /> :
                    <Image resize Mode="contain" style={styles.cardImage} source={require("../../assets/avatar.png")} />}
            </View>);
    }
    const services = [
        {
            id: "1",
            name: "Upload Image",
            action: (onPress) => {
                pickImage();
            }
        },
        {
            id: "2",
            name: "Update",
            action: (onPress) => {
                handleUpdates();
            }
        },
        {
            id: "3",
            name: "Go Back",
            action: (onPress) => setcurrentView({ screen: keys.screens.PROFILE, header: true, footer: true })
        }
    ]
    const _renderItem = (item) => {
        return <View style={styles.submitButton}
            key={item.id}>
            <TouchableOpacity
                onPress={() => item.action()} // to be added
                style={styles.listItemTouchable}>
                <View style={{ flex: 0.6 }}>
                    <Text style={styles.ServiceNameStyle}>{item.name}</Text>
                </View>
            </TouchableOpacity>
        </View>
    }
    return (
        <View style={styles.homeScreen}>
            <AvatarCard />
            <Text
                style={styles.EditProfileTitle}>
                Edit Profile
            </Text>
            <TextInput style={styles.input}
                underlineColorAndroid="transparent"
                value={userDetails.name}
                onChangeText={(text) => handleName(text)}
            />
            <TextInput style={styles.input}
                underlineColorAndroid="transparent"
                value={userDetails.email}
                onChangeText={(text) => handleEmail(text)}
            />
            <SafeAreaView style={{ flex: 1, marginBottom: 15 }}>
                {
                    services.map(_renderItem)
                }
            </SafeAreaView>
        </View>
    )
}

const styles = StyleSheet.create({
    homeScreen: {
        backgroundColor: keys.colors.HOMEBACKGROUND,
        flex: 1,
        width: '100%'
    },
    EditProfileTitle: {
        textAlign: 'center',
        fontWeight: 'bold',
        fontSize: 25,
        marginBottom: 10,
        color: keys.colors.MAIN
    },
    input: {
        backgroundColor: 'white',
        width: '92%',
        height: undefined,
        aspectRatio: 1 / 0.185,
        marginHorizontal: '4%',
        marginVertical: '1%',
        borderRadius: 5,
        textAlign: 'center',
        fontSize: 20
    },
    card: {
        backgroundColor: "white",
        width: '92%',
        height: "25%",
        borderStyle: 'solid',
        borderWidth: 1,
        marginHorizontal: '4%',
        marginVertical: '1.5%',
        borderRadius: 17,
        marginVertical: "2%",
        marginBottom: "8%",
        marginTop: "5%",
        flexDirection: 'row',
        justifyContent: "space-evenly",
        alignItems: 'center'
    },
    cardImage: {
        width: 120,
        height: 120,
        borderRadius: 150 / 2,
        overflow: "hidden",
        borderWidth: 3,
        borderColor: keys.colors.MAIN,
    },
    cardText: {
        textAlign: 'left',
        textAlignVertical: 'center',
        fontSize: 18,
        color: keys.colors.MAIN,
        padding: "2%",
        marginHorizontal: "2%",
    },
    submitButton: {
        backgroundColor: 'white',
        width: '92%',
        height: undefined,
        aspectRatio: 1 / 0.185,
        marginHorizontal: '4%',
        marginVertical: '1%',
        borderRadius: 5
    },
    listItemTouchable: {
        flexDirection: 'row',
        justifyContent: "space-evenly"
    },
    ServiceNameStyle: {
        fontSize: 20,
        alignSelf: 'center',
        margin: '10%'
    },
});
export default EditProfile