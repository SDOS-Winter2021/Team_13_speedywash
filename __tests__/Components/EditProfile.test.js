import React from 'react';
import renderer from 'react-test-renderer';
import EditProfile from "../../Components/Profile/EditProfile"
import {test_user} from "../global"

it('EditProfile renders correctly', () => {
    const tree = renderer.create(<EditProfile currentUser={test_user} />).toJSON();
    expect(tree).toMatchSnapshot();
})