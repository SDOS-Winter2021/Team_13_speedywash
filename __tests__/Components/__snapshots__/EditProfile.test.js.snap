// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfile renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#DCF2FD",
      "flex": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 17,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "height": "25%",
        "justifyContent": "space-evenly",
        "marginBottom": "8%",
        "marginHorizontal": "4%",
        "marginTop": "5%",
        "marginVertical": "2%",
        "width": "92%",
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#006BAD",
            "fontSize": 18,
            "marginHorizontal": "2%",
            "padding": "2%",
            "textAlign": "left",
            "textAlignVertical": "center",
          }
        }
      >
        test
      </Text>
      <Text
        style={
          Object {
            "color": "#006BAD",
            "fontSize": 18,
            "marginHorizontal": "2%",
            "padding": "2%",
            "textAlign": "left",
            "textAlignVertical": "center",
          }
        }
      >
        +919958820305
      </Text>
    </View>
    <Image
      Mode="contain"
      resize={true}
      source={
        Object {
          "uri": "https://firebasestorage.googleapis.com/v0/b/speedywash-7d67a.appspot.com/o/profile%2FxD1ncicw8lbnZGk9sTxPH9yCEi13?alt=media&token=85833276-6e3d-4a86-bbab-1d18c643d48a",
        }
      }
      style={
        Object {
          "borderColor": "#006BAD",
          "borderRadius": 75,
          "borderWidth": 3,
          "height": 120,
          "overflow": "hidden",
          "width": 120,
        }
      }
    />
  </View>
  <Text
    style={
      Object {
        "color": "#006BAD",
        "fontSize": 25,
        "fontWeight": "bold",
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    Edit Profile
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Object {
        "aspectRatio": 5.405405405405405,
        "backgroundColor": "white",
        "borderRadius": 5,
        "fontSize": 20,
        "height": undefined,
        "marginHorizontal": "4%",
        "marginVertical": "1%",
        "textAlign": "center",
        "width": "92%",
      }
    }
    underlineColorAndroid="transparent"
    value="test"
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Object {
        "aspectRatio": 5.405405405405405,
        "backgroundColor": "white",
        "borderRadius": 5,
        "fontSize": 20,
        "height": undefined,
        "marginHorizontal": "4%",
        "marginVertical": "1%",
        "textAlign": "center",
        "width": "92%",
      }
    }
    underlineColorAndroid="transparent"
    value="hardik18391@iiitd.ac.in"
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
        "marginBottom": 15,
      }
    }
  >
    <View
      style={
        Object {
          "aspectRatio": 5.405405405405405,
          "backgroundColor": "white",
          "borderRadius": 5,
          "height": undefined,
          "marginHorizontal": "4%",
          "marginVertical": "1%",
          "width": "92%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-evenly",
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 0.6,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "fontSize": 20,
                "margin": "10%",
              }
            }
          >
            Upload Image
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "aspectRatio": 5.405405405405405,
          "backgroundColor": "white",
          "borderRadius": 5,
          "height": undefined,
          "marginHorizontal": "4%",
          "marginVertical": "1%",
          "width": "92%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-evenly",
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 0.6,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "fontSize": 20,
                "margin": "10%",
              }
            }
          >
            Update
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "aspectRatio": 5.405405405405405,
          "backgroundColor": "white",
          "borderRadius": 5,
          "height": undefined,
          "marginHorizontal": "4%",
          "marginVertical": "1%",
          "width": "92%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-evenly",
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 0.6,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "fontSize": 20,
                "margin": "10%",
              }
            }
          >
            Go Back
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
